esphome/components/scd30/scd30.cpp

```diff
@@ -0,0 +1,177 @@
+static const uint16_t SCD30_CMD_AUTOMATIC_SELF_CALIBRATION = 0x5306;
+static const uint16_t SCD30_CMD_FORCED_CALIBRATION = 0x5204;
+static const uint16_t SCD30_CMD_TEMPERATURE_OFFSET = 0x5403;
+static const uint16_t SCD30_CMD_ALTITUDE_COMPENSATION = 0x5102;
+static const uint16_t SCD30_CMD_SOFT_RESET = 0xD304;
+
+void SCD30Component::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up scd30...");
+
+  /// Firmware version identification
+  if (!this->write_command_(SCD30_CMD_GET_FIRMWARE_VERSION)) {
+    this->error_code_ = COMMUNICATION_FAILED;
+    this->mark_failed();
+    return;
+  }
+  uint16_t raw_firmware_version[3];
+
+  if (!this->read_data_(raw_firmware_version, 3)) {
+    this->error_code_ = FIRMWARE_IDENTIFICATION_FAILED;
+    this->mark_failed();
+    return;
+  }
+  ESP_LOGD(TAG, "SCD30 Firmware v%0d.%02d", (uint16_t(raw_firmware_version[0]) >> 8),
+           uint16_t(raw_firmware_version[0] & 0xFF));
+
+  /// Sensor initialization
+  if (!this->write_command_(SCD30_CMD_START_CONTINUOUS_MEASUREMENTS)) {
+    ESP_LOGE(TAG, "Sensor SCD30 error starting continuous measurements.");
+    this->error_code_ = MEASUREMENT_INIT_FAILED;
+    this->mark_failed();
+    return;
+  }
+}
+
+void SCD30Component::dump_config() {
+  ESP_LOGCONFIG(TAG, "scd30:");
+  LOG_I2C_DEVICE(this);
+  if (this->is_failed()) {
+    switch (this->error_code_) {
+      case COMMUNICATION_FAILED:
+        ESP_LOGW(TAG, "Communication failed! Is the sensor connected?");
+        break;
+      case MEASUREMENT_INIT_FAILED:
+        ESP_LOGW(TAG, "Measurement Initialization failed!");
+        break;
+      case FIRMWARE_IDENTIFICATION_FAILED:
+        ESP_LOGW(TAG, "Unable to read sensor firmware version");
+        break;
+      default:
+        ESP_LOGW(TAG, "Unknown setup error!");
+        break;
+    }
+  }
+  LOG_UPDATE_INTERVAL(this);
+  LOG_SENSOR("  ", "CO2", this->co2_sensor_);
+  LOG_SENSOR("  ", "Temperature", this->temperature_sensor_);
+  LOG_SENSOR("  ", "Humidity", this->humidity_sensor_);
+}
+
+void SCD30Component::update() {
+  /// Check if measurement is ready before reading the value
+  if (!this->write_command_(SCD30_CMD_GET_DATA_READY_STATUS)) {
+    this->status_set_warning();
+    return;
+  }
+
+  uint16_t raw_read_status[1];
+  if (!this->read_data_(raw_read_status, 1) || raw_read_status[0] == 0x00) {
+    this->status_set_warning();
+    ESP_LOGW(TAG, "Data not ready yet!");
+    return;
+  } else {
+    if (!this->write_command_(SCD30_CMD_READ_MEASUREMENT)) {
+      ESP_LOGW(TAG, "Error reading measurement!");
+      this->status_set_warning();
+      return;
+    }
+
+    this->set_timeout(50, [this]() {
+      uint16_t raw_data[6];
+      if (!this->read_data_(raw_data, 6)) {
+        this->status_set_warning();
+        return;
+      }
+      unsigned int temp_c_o2_u32, temp_temp_u32, temp_hum_u32;
+      temp_c_o2_u32 = (unsigned int) ((((unsigned int) raw_data[0]) << 16) | ((unsigned int) raw_data[1]));
```

Comment:
```suggestion
      temp_c_o2_u32 = (((uint32_t(raw_data[0])) << 16) | (uint32_t(raw_data[1]));
```

Just a suggestion, but IMO much easier to read.

Reply:
I've refactored these lines, thanks.